Nav2 planner plugin tutorial (#1633)
ros-planning/navigation2: sphinx_doc/tutorials/docs/writing_new_nav2planner_plugin.rst

.. _writing_new_nav2planner_plugin:

Writing a new planner plugin
****************************

- `Overview`_
- `Requirements`_
- `Tutorial Steps`_

> .. image:: images/Writing_new_nav2planner_plugin/nav2_rrtconnect_gif.gif

Reviewer: Lets work on the planner before addressing videos specifically

    :width: 700px
    :align: center
    :alt: Animated gif with rrtconnect demo

Overview
========

This tutorial shows how to create you own planner plugin.

Requirements
============

- ROS2 (binary or build-from-source)
- Navigation2 (Including dependencies)
- Gazebo
- Turtlebot3

> - `OMPL (binary or build-from-source) <https://ompl.kavrakilab.org/installation.html>`_

Reviewer: Is there not some rosdep key for OMPL?

Author: Not sure, but need to check!

Author: OMPL has not been release for eloquent. I think we need to build from source as Movit2 guys does it for now.

Reviewer: I filed a ticket https://github.com/ros-gbp/ompl-release/issues/16

Ok, not sure what to do about this at this very moment, but this is a smaller thing and we'll figure it out.

Tutorial Steps
==============

1- Creating a new Planner Plugin
--------------------------------

We will create a Sampling base planner - RRT connect using OMPL library.
The annotated code in this tutorial can be found in `navigation_tutorials <https://github.com/ros-planning/navigation2_tutorials>`_ repository as the ``nav2_rrtconnect_plugin``
This package can be a considered as a reference for writing planner plugin.

> So, the class ``nav2_rrtconnect_planner::RRTConnect`` example plugin inherites from base class ``nav2_core::GlobalPlanner``. The base class provides 5 pure virtual methods API to implement a planner plugin in Navigation 2. The implemented plugin will be used by the planner server which is of `LifecycleNode <https://design.ros2.org/articles/node_lifecycle.html>`_ type.

Reviewer: ```suggestion
Our example plugin inherits from the base class ``nav2_core::GlobalPlanner``. The base class provides 5 pure virtual methods to implement a planner plugin. The plugin will be used by the planner server to compute trajectories.
```

> Lets see and learn more about these virtual methods which we have to override in order to write a planner plugin.

Reviewer: ```suggestion
Lets learn more about the methods needed to write a planner plugin.
```

| cleanup()            | Method is called when planner server goes to on_cleanup state. Ideally this| Yes                     |
|                      | method should clean up resoures which are created for the planner.         |                         |
+----------------------+----------------------------------------------------------------------------+-------------------------+
| createPlan()         | Method is called when planner server demands a global plan for specified   | Yes                     |
|                      | start and goal pose. This method returns `nav_msgs::msg::Path` carrying    |                         |
|                      | global plan. This method takes 2 input parmas: start pose and goal pose.   |                         |
+----------------------+----------------------------------------------------------------------------+-------------------------+

For this tutorial, we will be using methods ``RRTConnect::configure()`` and ``RRTConnect::createPlan()`` to create RRT Connect planner.

In our planner, ``RRTConnect::configure()`` method sets member variables and ROS parameters,

> .. code-block:: c

Reviewer: Small, but this is c++ not C

  node_ = parent;
  tf_ = tf;
  name_ = name;
  costmap_ = costmap_ros->getCostmap();
  global_frame_ = costmap_ros->getGlobalFrameID();

  // Parameter initialization
  nav2_util::declare_parameter_if_not_declared(node_, name_ + ".solve_time", rclcpp::ParameterValue(1.0));
  node_->get_parameter(name_ + ".solve_time", solve_time_);
  nav2_util::declare_parameter_if_not_declared(node_, name_ + ".max_trials", rclcpp::ParameterValue(4));
  node_->get_parameter(name_ + ".max_trials", max_trials_);
  nav2_util::declare_parameter_if_not_declared(node_, name_ + ".collision_checking_resolution", rclcpp::ParameterValue(0.001));
  node_->get_parameter(name_ + ".collision_checking_resolution", collision_checking_resolution_);
  nav2_util::declare_parameter_if_not_declared(node_, name_ + ".allow_unknown", rclcpp::ParameterValue(true));
  node_->get_parameter(name_ + ".allow_unknown", allow_unknown_);

> and also, it sets OMPL variables such as bounds, setup settings and space validation.

Reviewer: You should go over that too here

> In ``RRTConnect::createPlan()`` method, RRT Connect planner is called using start pose and goal pose to solve the global path planning problem. Upon succeding, it converts the path to the ``nav_msgs::msg::Path`` and returns to the caller which is planner server instance.

Reviewer: Please show this code as well.

> The remaining methods are not used but its mandatory to override them. As per the rules, we did override all but left them blank.

Reviewer: We shouldn't be encouraging people to ignore lifecycle states. Please implement them and then explain what they are doing. Keep in mind this is the first introduction into navigation planning for alot of folks so we want them to have good habits

Author: How would this be different from the explanation from the table?

Reviewer: ... good point.

2- Exporting the planner plugin
-------------------------------

Now that we have created our custom planner, we need to export our planner plugin so that it would be visible to the planner server. Plugins are loaded at runtime and if they are not visible, then our planner server won't be able to load it. In ROS2, exporting and loading plugins is handled by ``pluginlib``.

> Coming to our tutorial, class ``nav2_rrtconnect_planner::RRTConnect`` is loaded dynamically as ``nav2_core::GlobalPlanner`` which is our base class.

Reviewer: I'm not sure what you mean by loaded dynamically? at this point, you have some code in a file but there's no server. I think you just mean to point out that the planner derives from the GlobalPlanner class as a plugin interface.

Author: Yeah. I think this line is redundant. I should remove it.

1. To export the planner, we need to provide two lines

.. code-block:: c
  
  #include "pluginlib/class_list_macros.hpp"
  PLUGINLIB_EXPORT_CLASS(nav2_rrtconnect_planner::RRTConnect, nav2_core::GlobalPlanner)

Note that it requires pluginlib to export out plugin's class. Pluginlib would provide as macro ``PLUGINLIB_EXPORT_CLASS`` which does all the work of exporting.

It is good practice to place this lines at the end of the file but in practice you can also write at the top.

> 2. Next step would be to create plugin's description file in the root directory of the package. For example, ``global_planner_plugin.xml`` file in our tutorial package. This file contains following information

Reviewer: Is this information in the costmap tutorial? If not, can you add it? Its good context and this section is similar for both so ideally, I'd like them to be basically identical, minus the context / names.

Author: Sure!

 - ``library path``: Plugin's library name and it's location.
 - ``class name``: Name of the class.
 - ``class type``: Type of class.
 - ``base class``: Name of the base class.
 - ``description``: Description of the plugin.

.. code-block:: xml

  <library path="nav2_rrtconnect_planner_plugin">
    <class name="nav2_rrtconnect_planner/RRTConnect" type="nav2_rrtconnect_planner::RRTConnect" base_class_type="nav2_core::GlobalPlanner">
      <description>This is an example plugin which produces global path utilizing OMPL's implemented RRT Connect Algorithm.</description>
    </class>
  </library>

3. Next step would be to export plugin using ``CMakeLists.txt`` by using cmake function ``pluginlib_export_plugin_description_file()``. This function installs plugin description file to ``share`` directory and sets ament indexes to make it discoverable.

.. code-block:: text

  pluginlib_export_plugin_description_file(nav2_core global_planner_plugin.xml)

4. Plugin description file should also be added to ``package.xml``

.. code-block:: xml

  <export>
    <build_type>ament_cmake</build_type>
    <nav2_core plugin="${prefix}/global_planner_plugin.xml" />
  </export>

> 5. Put the package in a ROS2 workspace and compile. 

Reviewer: ```suggestion
5. Compile and it should be registered. Next, we'll use this plugin.
```

Reviewer: The hope is that this is already in a workspace

3- Pass the plugin name through params file
-------------------------------------------

> To enable the plugin, we need to modify the ``nav2_params.yml`` file as below

Reviewer: `.yaml`

replace following params

.. code-block:: text

  planner_server:
  ros__parameters:
    planner_plugin_types: ["nav2_navfn_planner/NavfnPlanner"]
    planner_plugin_ids: ["GridBased"]
    use_sim_time: True
    GridBased.tolerance: 2.0
    GridBased.use_astar: false
    GridBased.allow_unknown: true

with

.. code-block:: text

  planner_server:
  ros__parameters:

>     planner_plugin_types: ["nav2_rrtconnect_planner/RRTConnect"]

Reviewer: Missing planner plugins IDs

    use_sim_time: True
    GridBased.solve_time: 1.0
    GridBased.max_trials: 4
    GridBased.collision_checking_resolution: 0.001

>     GridBased.allow_unknown: True

Reviewer: `true`

> 4- Run GradientLayer plugin

Reviewer: Gradient plugin :wink: ?

---------------------------

Run Turtlebot3 simulation with enabled navigation2. Detailed instruction how to make it are written at :ref:`getting_started`. Below is shortcut command for that:

.. code-block:: bash

  $ ros2 launch nav2_bringup tb3_simulation_launch.py params_file:=/path/to/your_params_file.yaml

Then goto RViz and click on the "2D Pose Estimate" button at the top and point the location on map as it was described in :ref:`getting_started`. Robot will localize on the map and then click on "Navigation2 goal" and click on the pose where you want your planner to consider a goal pose. After that planner will plan the path and robot will start moving towards the goal. Please checkout the video below for demo.

.. raw:: html

    <h1 align="center">
      <div>
        <div style="position: relative; padding-bottom: 0%; overflow: hidden; max-width: 100%; height: auto;">

>           <iframe width="450" height="300" src="https://www.youtube.com/embed/MSy7WAE_xz4?autoplay=1&mute=1" frameborder="1" allowfullscreen></iframe>

Reviewer: Some of these paths are _really_ bad -  I think we need to do some work on the planner itself.

I think there's 2 ways to make the tutorial. A toy example (straight line planner, something unrealistic for actual use so no one actually tries to) or a working example (you can use it if you want since it works, but simple). We need to find a balance to put this in. I think it would look really bad on ROS2 Navigation if folks wanted to use a sampling based planner and saw these results as the "out of the box" experience with them.

Reviewer: If you'd like, we could make a straight line planner as the tutorial and work together on refining this OMPL work into a supported planner in navigation.

Author: I don't mind if you are fine with it. I was thinking of making tutorials with some thing concrete where the user gets something that will work in the end.

.. _writing_new_nav2planner_plugin:

Writing a new planner plugin
****************************

- `Overview`_
- `Requirements`_
- `Tutorial Steps`_

Overview
========

This tutorial shows how to create you own planner plugin.

Requirements
============

- ROS2 (binary or build-from-source)
- Navigation2 (Including dependencies)
- Gazebo
- Turtlebot3

Tutorial Steps
==============

1- Creating a new Planner Plugin
--------------------------------

We will create a simple straight line planner.

> The annotated code in this tutorial can be found in `navigation_tutorials <https://github.com/ros-planning/navigation2_tutorials>`_ repository as the ``nav2_rrtconnect_plugin``

Reviewer: You should rename `nav2_rrtconnect_plugin` -> `nav2_straightline_planner` or whatever you'd like to name it

+----------------------+----------------------------------------------------------------------------+-------------------------+
| cleanup()            | Method is called when planner server goes to on_cleanup state. Ideally this| Yes                     |
|                      | method should clean up resoures which are created for the planner.         |                         |
+----------------------+----------------------------------------------------------------------------+-------------------------+
| createPlan()         | Method is called when planner server demands a global plan for specified   | Yes                     |
|                      | start and goal pose. This method returns `nav_msgs::msg::Path` carrying    |                         |
|                      | global plan. This method takes 2 input parmas: start pose and goal pose.   |                         |
+----------------------+----------------------------------------------------------------------------+-------------------------+

For this tutorial, we will be using methods ``StraightLine::configure()`` and ``StraightLine::createPlan()`` to create straight line planner.

> In our planner, ``StraightLine::configure()`` method sets member variables and ROS parameters,

Reviewer: ```suggestion
In planners, ``configure()`` method must set member variables from ROS parameters and any initialization required,
```

.. code-block:: c++

  node_ = parent;
  tf_ = tf;
  name_ = name;
  costmap_ = costmap_ros->getCostmap();
  global_frame_ = costmap_ros->getGlobalFrameID();

  // Parameter initialization
  nav2_util::declare_parameter_if_not_declared(node_, name_ + ".interpolation_resolution", rclcpp::ParameterValue(0.1));
  node_->get_parameter(name_ + ".interpolation_resolution", interpolation_resolution_);

> In ``StraightLine::createPlan()`` method, straight line planner is called using start pose and goal pose to solve the global path planning problem. Upon succeding, it converts the path to the ``nav_msgs::msg::Path`` and returns to the caller which is planner server instance. Below annotation shows the implementation of this method.

Reviewer: ```suggestion
In ``createPlan()`` method, we need to create a path from the given start to goal poses. The ``StraightLine::createPlan()`` is called using start pose and goal pose to solve the global path planning problem. Upon succeeding, it converts the path to the ``nav_msgs::msg::Path`` and returns to the planner server. Below annotation shows the implementation of this method.
```

.. code-block:: c++

  nav_msgs::msg::Path global_path;

  // Checking if the goal and start state is in the global frame
  if (start.header.frame_id != global_frame_) {
    RCLCPP_ERROR(
      node_->get_logger(), "Planner will only except start position from %s frame",
      global_frame_.c_str());
    return global_path;
  }

  if (goal.header.frame_id != global_frame_) {
    RCLCPP_INFO(
      node_->get_logger(), "Planner will only except goal position from %s frame",
      global_frame_.c_str());
    return global_path;
  }

  global_path.poses.clear();
  global_path.header.stamp = node_->now();
  global_path.header.frame_id = global_frame_;
  global_path.poses.push_back(start);
  // calculating the number of loops for current value of interpolation_resolution_

>   int total_number_of_loop = std::abs(goal.pose.position.x - start.pose.position.x)/interpolation_resolution_;

Reviewer: This should probably be total distance, e.g. (`hypot(delta x, delta y)`) and then interpolating along the total distance, this would have very different results if the path was in a straight line along the Y axis (e.g. number of loops = 0).

+----------------------+----------------------------------------------------------------------------+-------------------------+
| createPlan()         | Method is called when planner server demands a global plan for specified   | Yes                     |
|                      | start and goal pose. This method returns `nav_msgs::msg::Path` carrying    |                         |
|                      | global plan. This method takes 2 input parmas: start pose and goal pose.   |                         |
+----------------------+----------------------------------------------------------------------------+-------------------------+

For this tutorial, we will be using methods ``StraightLine::configure()`` and ``StraightLine::createPlan()`` to create straight line planner.

In planners, ``configure()`` method must set member variables from ROS parameters and any initialization required,

.. code-block:: c++

  node_ = parent;
  tf_ = tf;
  name_ = name;
  costmap_ = costmap_ros->getCostmap();
  global_frame_ = costmap_ros->getGlobalFrameID();

  // Parameter initialization

>   nav2_util::declare_parameter_if_not_declared(node_, name_ + ".interpolation_resolution", rclcpp::ParameterValue(0.1));

Reviewer: Oh! Oh! One more thing, please example the `name.<param>` thing. That is going to trip a bunch of people up. Explaining that the plugin is mapped to the plugin ID so you can have multiple plugins loaded of the same type. So the `name_.` is used for getting the parameters namespaced to the right version.

Author: okay cool!
